### PR TITLE
Remove unnecessary string cloning from the parser

### PR DESCRIPTION
**Pull request by charliermarsh**
Closes https://github.com/astral-sh/ruff/issues/9869.

### CODE REVIEW

**Comment by charliermarsh** [comment]
Hate this change, didn't even improve benchmarks!!!

**Comment by MichaReiser** [comment]
> Hate this change, didn't even improve benchmarks!!!

That's because codspeed uses the cost 0 for allocations....

File: crates/ruff_python_parser/src/string.rs [hunk 1]
```diff
@@ -227,77 +254,160 @@ impl<'a> StringParser<'a> {
...
-                        return Err(LexicalError::new(
-                            LexicalErrorType::OtherError(
-                                "bytes can only contain ASCII literal characters".to_string(),
-                            ),
-                            self.get_pos(),
-                        ));
-                    }
-                    content.push(ch);
+    fn parse_bytes(mut self) -> Result<StringType, LexicalError> {
+        let index = first_non_ascii_byte(self.source.as_bytes());
```
Comment (charliermarsh) [inline crates/ruff_python_parser/src/string.rs:291]: @BurntSushi - Is this exposed in any crates that we already use? I had to vendor it from `bstr`.

Comment (BurntSushi) [reply crates/ruff_python_parser/src/string.rs:291]: Yes. :P https://docs.rs/bstr/latest/bstr/trait.ByteSlice.html#method.find_non_ascii_byte

**Review by dhruvmanila — APPROVED** [review]
Nice!

**Review by MichaReiser — APPROVED** [review]

**Review by MichaReiser — APPROVED** [review]
:guitar:

**Comment by charliermarsh** [comment]
I'm gonna get to the bottom of this if it's the last thing I do.

**Comment by charliermarsh** [comment]
I see basically no improvement on local benchmarks either via criterion or hyperfine. I'm not sure why, to be honest. I guess I'll proceed with the change since it's both (1) a better API, and (2) intuitively better to avoid re-allocating here when we don't need to. But I'm a bit confused.

**Comment by charliermarsh** [comment]
(Ignore current state of the PR, I duplicated the existing and updated code to do some benchmarking. I'll clean up later.)

**Comment by charliermarsh** [comment]
By removing lexing from the parser benchmark, I'm now able to see a more consistent 1-2% improvement in parse time, so I'll move forward with this. (Benchmarking the individual string-parsing functions (rather than the parser as a whole) shows an enormous speed-up, like 2x.)

**Pull request merged**